feat(admin): add basic true pagination functionality to admin users table

mikais13 opened the pull request:
# Description

<!-- Please include a summary of the changes towards the related issue. Please also  include relevant context. List any dependencies that are required for this change. -->

Add a pagination specific admin table version that takes in a `useInfiniteQuery` style hook and fetches relevant data through the pagination, rather then supplying the table with the full dataset.
Add true pagination functionality to the users admin table using tanstack query using this component.

Note that the filtering and searching of data doesn't work properly currently as these are implemented internally within the table and can only access fetched data, therefore can't search/filter for data that hasn't been loaded in by opening the relevant page.

Fixes #755 

## Type of change
<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

https://github.com/user-attachments/assets/a8a02557-017b-439f-83a8-ab474a807bcd

- [x] Manual testing (requires screenshots or videos)
- [ ] Integration tests written (requires checks to pass)

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added thorough tests that prove my fix is effective and that my feature works
- [x] I've requested a review from another user

taroj1205 commented on apps/frontend/src/services/auth/useUpdateSelfMutation.ts:
```diff
@@ -0,0 +1,24 @@
+      const parsed = UpdateSelfRequestSchema.parse(data)
+      return await AuthService.patchMe(parsed, token)
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ["auth", "me", token] })
+    },
```
can we also cancelQueries?
also please remove `token` from `queryKey`

taroj1205 commented on apps/frontend/src/services/admin/semester/AdminSemesterQueries.ts:
```diff
@@ -0,0 +1,54 @@
+    mutationFn: AdminSemesterService.createSemester,
+    // TODO: when get by id is implemented, do not invalidate get by id queries
+    onSuccess: () =>
+      queryClient.invalidateQueries({
+        queryKey: [QueryKeys.SEMESTER_QUERY_KEY],
+      }),
```
can we also cancelQueries?

taroj1205 commented on apps/frontend/src/services/admin/semester/AdminSemesterQueries.ts:
```diff
@@ -0,0 +1,54 @@
+  return useMutation({
+    mutationFn: AdminSemesterService.updateSemester,
+    // TODO: when get by id is implemented, only invalidate the updated id for get by id or update the specific query with the updated data
+    onSuccess: () =>
+      queryClient.invalidateQueries({
+        queryKey: [QueryKeys.SEMESTER_QUERY_KEY],
+      }),
+  })
```
can we also cancelQueries?

taroj1205 commented on apps/frontend/src/services/admin/semester/AdminSemesterQueries.ts:
```diff
@@ -0,0 +1,54 @@
+  return useMutation({
+    mutationFn: AdminSemesterService.deleteSemester,
+    // TODO: when get by id is implemented, only invalidate the updated id for get by id
+    onSuccess: () =>
+      queryClient.invalidateQueries({
+        queryKey: [QueryKeys.SEMESTER_QUERY_KEY],
+      }),
+  })
```
can we also cancelQueries?

taroj1205 commented on packages/shared/src/utils/typeGuards.ts:
```diff
@@ -0,0 +1,21 @@
+import { Gender, PlayLevel } from "../types/enums"
+
+/**
+ * Checks if the provided value is a valid Gender enum value. Returns the value if valid, otherwise undefined.
+ */
+export const isGender = (value: unknown): Gender | undefined => {
+  if (typeof value !== "string") {
+    return undefined
+  }
+  return Object.values(Gender).includes(value as Gender) ? (value as Gender) : undefined
+}
+
+/**
+ * Checks if the provided value is a valid PlayLevel enum value. Returns the value if valid, otherwise undefined.
+ */
+export const isPlayLevel = (value: unknown): PlayLevel | undefined => {
+  if (typeof value !== "string") {
+    return undefined
+  }
+  return Object.values(PlayLevel).includes(value as PlayLevel) ? (value as PlayLevel) : undefined
```
we seem to already have `type-gards.ts`

taroj1205 commented on packages/ui/src/components/Primitive/Button/button.style.ts:
I think this file was already deleted

taroj1205 commented on packages/ui/src/components/Primitive/Button/icon-button.style.ts:
I think this file was already deleted

mikais13 commented:
Sorry guys, just fixed up all the weird file stuff, did not realise how bad it was until just now. I think that at some point when updating the branch, git used a corrupted version of the branch that had some old files or something in there. I think I have removed all of the corrupted stuff, but if there is anything I missed do let me know.

taroj1205 commented on packages/ui/src/components/Composite/AdminTable/AdminTableWithPaginatedQuery.test.tsx:
```diff
@@ -0,0 +1,346 @@
+  }
+})
+
+vi.mock("nuqs", () => ({
+  parseAsInteger: {
+    withDefault: (defaultValue: number) => ({ defaultValue }),
+  },
+  parseAsString: {
+    withDefault: (defaultValue: string) => ({ defaultValue }),
+  },
+  parseAsArrayOf: (_parser: unknown) => ({
+    withDefault: (defaultValue: unknown[]) => ({ defaultValue }),
+  }),
+  useQueryStates: () => [{ page: currentPage, perPage: 20 }, mockSetState],
+}))
```
you should be able to just use the nuqs testing adapter
https://github.com/UoaWDCC/uabc-web/blob/014263b2324a6caab08f8766b99e3e0f2abe9eb1/packages/ui/src/components/Generic/ManagementTable/MemberManagementContext.test.tsx#L4

taroj1205 commented on packages/ui/src/components/Composite/AdminTable/AdminTableWithPaginatedQuery.tsx:
```diff
@@ -0,0 +1,245 @@
+      async (targetPage: number) => {
+        const pagesNeeded = targetPage - currentPageCount
+        if (pagesNeeded > 0 && hasNextPage && !isFetchingPages) {
+          setIsFetchingPages(true)
+          try {
+            for (let i = 0; i < pagesNeeded; i++) {
+              if (hasNextPage) {
+                await fetchNextPage()
+              }
+            }
+          } finally {
+            setIsFetchingPages(false)
+          }
+        }
+      },
```
could this use Promise.all maybe? i'm not sure

taroj1205 replied:
ah thats true

taroj1205 reviewed (commented):
sorry i missed something in my previous review

jeffplays2005 reviewed (approved):
LGTM!
